src/Symfony/Component/Security/Http/Firewall/ContextListener.php

```diff
@@ -288,15 +288,23 @@ private static function hasUserChanged(UserInterface $originalUser, TokenInterfa
         }
 
         if ($originalUser instanceof PasswordAuthenticatedUserInterface || $refreshedUser instanceof PasswordAuthenticatedUserInterface) {
-            if (!$originalUser instanceof PasswordAuthenticatedUserInterface || !$refreshedUser instanceof PasswordAuthenticatedUserInterface || $originalUser->getPassword() !== $refreshedUser->getPassword()) {
+            if (!$originalUser instanceof PasswordAuthenticatedUserInterface || !$refreshedUser instanceof PasswordAuthenticatedUserInterface) {
                 return true;
             }
 
+            if (null !== ($originalPassword = $originalUser->getPassword())
+                && ($refreshedPassword = $refreshedUser->getPassword()) !== $originalPassword
+                && (16 !== \strlen($originalPassword) || hash('xxh3', $refreshedPassword) !== $originalPassword)
+            ) {
+                return true;
+            }
+
+
             if ($originalUser instanceof LegacyPasswordAuthenticatedUserInterface xor $refreshedUser instanceof LegacyPasswordAuthenticatedUserInterface) {
                 return true;
             }
 
-            if ($originalUser instanceof LegacyPasswordAuthenticatedUserInterface && $refreshedUser instanceof LegacyPasswordAuthenticatedUserInterface && $originalUser->getSalt() !== $refreshedUser->getSalt()) {
+            if ($originalUser instanceof LegacyPasswordAuthenticatedUserInterface && $originalUser->getSalt() !== $refreshedUser->getSalt()) {
```

Comment:
the removed condition is redundant with the previous check, sorry psalm if you cannot get it